File: src/simulation/culture/aspect/dependency/AbstractDependency.java
package simulation.culture.aspect.dependency;

import simulation.space.resource.tag.ResourceTag;

import java.util.Objects;

abstract class AbstractDependency implements Dependency {
    protected ResourceTag tag;

    AbstractDependency(ResourceTag tag) {
        this.tag = tag;
    }

    @Override
    public boolean isPhony() {
        return tag.name.equals("phony");
    }

    @Override
    public ResourceTag getType() {
        return tag;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AbstractDependency that = (AbstractDependency) o;
        return Objects.equals(tag, that.tag);
    }

    @Override
    public int hashCode() {
        return Objects.hash(tag);
    }
}

File: src/simulation/culture/aspect/dependency/AbstractDependency.kt
package simulation.culture.aspect.dependency

import java.util.*

abstract class AbstractDependency(override val isPhony: Boolean) : Dependency {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as AbstractDependency
        return isPhony == that.isPhony
    }

    override fun hashCode(): Int {
        return Objects.hash(isPhony)
    }
}